contracts/TaskBook.sol

```diff
@@ -0,0 +1,89 @@
+pragma solidity ^0.4.4;
+
+import './AccountManager.sol';
+
+contract TaskBook is AccountManager {
+
+	uint private numTasks = 0;
+	uint constant maxSolvers = 10;
+	uint constant maxChallengers = 10;
+
+	event NewTask(uint taskID, uint minDeposit, uint blockNumber);
+	event SolveTask(uint indexed taskID, address solver, bytes32 taskData, uint minDeposit);
+	event SubmitSolution(uint taskID, bytes32 solution, uint minDeposit, bytes32 taskData);
+
+	mapping(uint => Task) private tasks;
+	mapping(uint => bytes32) private solverRandomBits;
+	mapping(uint => mapping(address => uint)) private challengers;
+
+	struct Task {
+		address owner;
+		address[] solvers;
+		address selectedSolver;
+		uint minDeposit;
+		bytes32 taskData;
+		uint numSolvers;
+		address[] challengers;
+		uint numChallengers;
+	}
+
+	//Task Issuers create tasks to be solved
+	function newTask(uint minDeposit, bytes32 taskData, uint numBlocks) returns (bool) {
+		require(balances[msg.sender] >= minDeposit);
+		tasks[numTasks] = Task(msg.sender, new address[](maxSolvers), 0x0, minDeposit, taskData, 0, new address[](maxSolvers), 0);
+		log0(sha3(msg.sender));//possible bug if log is after event
+		NewTask(numTasks, minDeposit, block.number+numBlocks);
+		numTasks++;
+		return true;
+	}
+
+	//Solver registers for tasks
+	function registerForTask(uint id, uint minDeposit) returns(bool) {
+		require(balances[msg.sender] >= minDeposit);
+		Task t = tasks[id];
+		require(!(t.owner == 0x0));
+		require(t.numSolvers < maxSolvers);
+		//random = sha3(random, block.blockhash(block.number-1));
+		t.solvers[t.numSolvers] = msg.sender;
+		t.numSolvers++;
+		log0(bytes32(sha3(msg.sender)));
+		return true;
+	}
+
+	//Task Issuer tells TrueBitExchange to select a solver
+	function selectSolver(uint taskID) returns (bool) {
+		Task t = tasks[taskID];
+		require(msg.sender == t.owner);
+		//uint random = uint(block.blockhash(block.number-1));
+		//address randomSolver = t.solvers[uint(random) % t.numSolvers];
+		address solver = t.solvers[0];
+		t.selectedSolver = solver;
+		SolveTask(taskID, solver, t.taskData, t.minDeposit);
+		return true;
+	}
+
+	//Selected solver submits a solution to the exchange
+	function submitSolution(uint taskID, bytes32 randomBits, bytes32 correctSolutionHash, bytes32 incorrectSolutionHash) returns (bool) {
+		Task t = tasks[taskID];
+		require(t.selectedSolver == msg.sender);
+		solverRandomBits[taskID] = sha3(randomBits);//save for later use
+		uint randomNum = uint(sha3(randomBits, block.blockhash(block.number-1)));
+		if (randomNum % 2 == 0) {
```

Comment:
I thought there was a two-step process here; the solver needs to commit to both possible solutions (one correct, one incorrect) and then once they know whether a forced error is in effect (after the next block, by hashing it on their own side), they choose which solution to submit as the "correct" one. Following that, making the first solution consistently correct and the second consistently incorrect would give away whether or not a forced error is in effect to the validators.

With the code as it is, I as a validator know whether a forced error is in effect or not by replaying this transaction, but I shouldn't know that until I've challenged the solver's submitted solution.

Reply:
Good point. Breaking it up into two steps seems reasonable. Not sure how that would look though.

At what point would they commit the solutions, and then upload the random bits? Perhaps this will be clearer when we incorporate the verification game part more.